Composer/packages/form-dialogs/src/FormDialogSchemaEditor.tsx

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import * as React from 'react';
+// eslint-disable-next-line @typescript-eslint/camelcase
+import { RecoilRoot, useRecoilTransactionObserver_UNSTABLE } from 'recoil';
+import { formDialogSchemaJsonSelector } from 'src/atoms/appState';
+import { useHandlers } from 'src/atoms/handlers';
+import { VisualEditor } from 'src/components/VisualEditor';
+
+export type FormDialogSchemaEditorProps = {
+  /**
+   * Unique id for the visual editor.
+   */
+  editorId: string;
+  /**
+   * Wether to show or hide the theme picker.
+   */
+  showThemePicker?: boolean;
+  /**
+   * Initial json schema content.
+   */
+  schema: { id: string; content: string };
+  /**
+   * Form dialog schema file extension.
+   */
+  schemaExtension?: string;
+  /**
+   * Record of available schema templates.
+   */
+  templates?: string[];
+  /**
+   * Callback for when the json schema update is updated.
+   */
+  onSchemaUpdated: (id: string, content: string) => void;
+  /**
+   * Callback for generating dialog using current valid form dialog schema.
+   */
+  onGenerateDialog: (formDialogSchemaJson: string) => void;
+};
+
+const InternalFormDialogSchemaEditor = React.memo((props: FormDialogSchemaEditorProps) => {
+  const {
+    editorId,
+    schema,
+    showThemePicker = false,
+    templates = [],
+    schemaExtension = '.schema',
+    onSchemaUpdated,
+    onGenerateDialog,
+  } = props;
+
+  const { setTemplates, reset, importSchemaString } = useHandlers();
+
+  React.useEffect(() => {
+    setTemplates({ templates });
+  }, [templates]);
+
+  React.useEffect(() => {
+    importSchemaString(schema);
+  }, [editorId]);
+
+  const startOver = React.useCallback(() => {
+    reset({ name: editorId });
+  }, [reset, editorId]);
+
+  useRecoilTransactionObserver_UNSTABLE(async ({ snapshot, previousSnapshot }) => {
+    const content = await snapshot.getPromise(formDialogSchemaJsonSelector);
+    const prevContent = await previousSnapshot.getPromise(formDialogSchemaJsonSelector);
+    if (content !== prevContent) {
```

Comment:
The getPromise is getting a snapshot of just the value of that one selector value.  I think this is the schema JSON string, so it ends up comparing two JSON strings to see if that has changed.  It isn't comparing two snapshot classes.  The comparison is necessary because the transaction observer fires for all the other changes going on in the app under the same RecoilRoot and it would cause the onSchemaUpdated to fire constantly without limiting the change detection to just the schema.

Reply:
It's comparing two string value of the schema in json.